rust/src/interpreter/utils.rs

```diff
@@ -0,0 +1,116 @@
+use std::cmp::min;
+
+use evmc_vm::{ExecutionMessage, MessageFlags, Revision, StatusCode};
+
+use crate::{interpreter::gas::consume_copy_cost, types::u256};
+
+pub trait SliceExt {
+    fn get_within_bounds(&self, offset: u256, len: u64) -> &[u8];
+
+    fn set_to_zero(&mut self);
+
+    fn copy_padded(&mut self, src: &[u8], gas_left: &mut u64) -> Result<(), StatusCode>;
+}
+
+impl SliceExt for [u8] {
+    #[inline(always)]
+    fn get_within_bounds(&self, offset: u256, len: u64) -> &[u8] {
+        if len == 0 {
+            return &[];
+        }
+        let (offset, offset_overflow) = offset.into_u64_with_overflow();
+        if offset_overflow {
+            return &[];
+        }
+        let offset = offset as usize;
+        let len = len as usize;
+        let (end, end_overflow) = offset.overflowing_add(len);
+        if end_overflow || offset >= self.len() {
+            &[]
+        } else {
+            &self[offset..min(end, self.len())]
+        }
+    }
+
+    #[inline(always)]
+    fn set_to_zero(&mut self) {
+        for byte in self {
+            *byte = 0;
+        }
+    }
+
+    #[inline(always)]
+    fn copy_padded(&mut self, src: &[u8], gas_left: &mut u64) -> Result<(), StatusCode> {
+        consume_copy_cost(gas_left, self.len() as u64)?;
+        self[..src.len()].copy_from_slice(src);
+        self[src.len()..].set_to_zero();
+        Ok(())
+    }
+}
+
+#[inline(always)]
+pub fn word_size(byte_len: u64) -> Result<u64, StatusCode> {
+    let (end, overflow) = byte_len.overflowing_add(31);
+    if overflow {
+        return Err(StatusCode::EVMC_OUT_OF_GAS);
+    }
+    Ok(end / 32)
+}
+
+#[inline(always)]
+pub(super) fn check_min_revision(
+    min_revision: Revision,
+    revision: Revision,
+) -> Result<(), StatusCode> {
+    if revision < min_revision {
+        return Err(StatusCode::EVMC_UNDEFINED_INSTRUCTION);
+    }
+    Ok(())
+}
+
+#[inline(always)]
+pub(super) fn check_not_read_only(
```

Comment:
This is a very surprising utility function. Where would this function be used? And is there any reuse?

Reply:
This is called for `SStore`, `TStore`, `SelfDestruct`, `Log*`, `Create`, `Create2`, `Call` and `CallCode` to fail if the read-only / static flag was set.